oasis/job.py

```diff
@@ -2,31 +2,71 @@
 
 from __future__ import absolute_import
 
+from threading import Thread
 from oasis.datasource import DataSource
+from oasis.libs.log import logger
+from oasis.models import Models
 
-DEFAULT_JOB_TIMEOUT = "240h"
+THREAD_JOIN_TIMEOUT = 5
 
 
 class Job(object):
     """Job defines config for calculation job."""
     def __init__(self, id, data_source,
-                 model, metrics, slack_channel,
-                 timeout=DEFAULT_JOB_TIMEOUT, config=None):
+                 models, slack_channel):
         self.id = id
         self.data_source = DataSource(data_source["url"])
-        self.model = model
-        self.metrics = metrics
+        self.models = models
         self.slack_channel = slack_channel
-        self.timeout = timeout
-        self.config = config
-    # def __init__(self, data):
-    # self.__dict__ = json.loads(data)
+        self.running_models = []
+        self.threads = dict()
 
     def __iter__(self):
         yield "id", self.id
         yield "data_source", dict(self.data_source)
-        yield "model", self.model
-        yield "metrics", self.metrics
+        yield "models", self.models
         yield "slack_channel", self.slack_channel
-        yield "timeout", self.timeout
-        # return self.__dict__
+
+    def run(self):
+        logger.info("[job:{job}] start to run"
+                    .format(job=dict(self)))
+        for md in self.models:
+            model = Models[md.get("name")]
+            model_runner = model(self.id, md, self.data_source, self.slack_channel)
+            t = Thread(target=model_runner.run())
+            t.start()
+            self.running_models.append(model_runner)
+            self.threads[md.get("name")] = t
+
+    def close(self):
+        logger.info("[job-id:{job_id}] start to stop"
+                    .format(job_id=self.id))
+        for model in self.running_models:
+            model.close()
+
+        for key in self.threads:
```

Comment:
key-> module?

Reply:
key is the name of the model

Reply:
so, name it as module?

Reply:
every model will run in one single thread, so, I add every model threads to the job for this job can close all model threads.  I don't understand what you mean.